sql: decompose a table descriptor into elements

fqazi opened the pull request:
Previously, we didn't have an proper element mapping
  for tables, views, and sequences inside the new schema
  changer. This was inadequate because we didn't have a
  proper decomposition making create operations difficult.
  To address this, this patch decomposes table descriptors
  into elements and uses them to implement DROP operations.

  Release justification: low risk and change to a disabled feature.
  Release note: None

fqazi committed:
sql: decompose a table descriptor into elements

Previously, we didn't have an proper element mapping
for tables, views, and sequences inside the new schema
changer. This was inadequate because we didn't have a
proper decomposition making create operations difficult.
To address this, this patch decomposes table descriptors
into elements and uses them to implement DROP operations.

Release justification: low risk and change to a disabled feature.
Release note: None

fqazi committed:
sql: add support for optimizing declarative schema changer plans

Previously, the declarative schema changer could potentially
execute unnecessary operations during drops. This was
inadequate because this led to extra work or impossible
operations (for example not all the columns of a table
can be dropped). To address this, this patch add supports
for eliminating unnecessary edges and emitting empty
stages.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/72856)
<!-- Reviewable:end -->

ajwerner commented on pkg/sql/schemachanger/rel/database.go:
```diff
@@ -103,6 +103,41 @@ func (t *Database) Insert(v interface{}) error {
 
+// Delete deletes an entity.
+func (t *Database) Delete(v interface{}) error {
+	e, err := toEntity(t.schema, v)
+	if err != nil {
+		return err
+	}
+	if err := t.delete(e); err != nil {
+		return err
+	}
+	for _, v := range e.m {
+		_, isEntity := t.schema.entityTypeSchemas[reflect.TypeOf(v)]
+		_, alreadyDefined := t.entities[v]
+		if isEntity && alreadyDefined {
+			if err := t.Delete(v); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (t *Database) delete(e *entity) error {
+	self := e.getComparableValue(t.schema, Self)
+
+	delete(t.entities, self)
+	for i := range t.indexes {
+		idx := &t.indexes[i]
+		idx.tree.Delete(&containerItem{
+			entity:    e,
+			indexSpec: &idx.indexSpec,
+		})
+	}
+	return nil
```
My first instict was that we should factor our a pattern here. Instead, though, I think we should see if this entity exists in the entities map. If it doesn't, just return nothing. If it does, iterate through its members and delete them rather than re-decomposing the entity.

ajwerner commented on pkg/sql/schemachanger/scpb/uml/table.puml:
```diff
@@ -0,0 +1,266 @@
+
+Column <|-- OnUpdateExpr
+
+object ComputedExpr
+
+OnUpdateExpr : TableID
+OnUpdateExpr : ColumnID
+OnUpdateExpr : []UsesSequenceID
+OnUpdateExpr : Expression
```
ComputedExpr fields?

ajwerner commented on pkg/sql/schemachanger/scpb/uml/table.puml:
```diff
@@ -0,0 +1,266 @@
+@startuml
```
Could we be generating this? What would it take? Could we use tags somewhere or maybe just the arrows somewhere?

ajwerner reviewed (commented):
Haven't digested the optimization parts yet. This is great. I'm sort of inclined to take another pass or two and then merge first and ask questions later.

fqazi committed:
sql: add automated UML diagram generation for new schema changer

Previously, the declarative schema changer had manually built
UML diagrams for the various elements. This was inadequate,
since there was extra effort keeping this data and the code
in sync. To address this, this patch adds support for automated
generation of this UML diagram. The dependencies between objects
are encoded as a struct tag inside the element protobuf.

Release note: None

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72856)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/sql/schemachanger/scpb/uml/table.puml, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72856#-MogW497Ai5l-8upJIxb-r2-1:-Moo9pTa9vY2X6ccMkdT:bhloupb) ([raw file](https://github.com/cockroachdb/cockroach/blob/364ecc62799f4104657e9ce6cf596cf54397926a/pkg/sql/schemachanger/scpb/uml/table.puml#L1)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Could we be generating this? What would it take? Could we use tags somewhere or maybe just the arrows somewhere?
</blockquote></details>

Done.

Added a generator for the diagram, where the arrows, for now, are manually encoded. The rest of it is done with reflection magic.
___
*[pkg/sql/schemachanger/scpb/uml/table.puml, line 71 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72856#-MogW497Ai5l-8upJIxb-r2-71:-Moo9q7A3ihyyLVx_vit:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/364ecc62799f4104657e9ce6cf596cf54397926a/pkg/sql/schemachanger/scpb/uml/table.puml#L71)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

ComputedExpr fields?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

postamar commented on pkg/sql/schemachanger/scexec/executor.go:
```diff
@@ -23,6 +23,12 @@ import (
+	// It is perfectly valid to have empty stage after optimizations /
+	// transformations.
+	if toExecute == nil {
+		log.Infof(ctx, "skipping execution no operations in this stage")
```
nit: "skipping execution, no operations in this stage" instead ?

postamar commented on pkg/sql/schemachanger/scexec/scmutationexec/scmutationexec.go:
```diff
@@ -262,19 +262,26 @@ func (m *visitor) AddTypeBackRef(ctx context.Context, op scop.AddTypeBackRef) er
 		}
 	}
-	if len(tbl.DependedOnBy) == 0 {
-		tbl.DependedOnBy = nil
```
I should have left a comment, but I put this nil-setting code here for a reason. See my comment about `[]` entries in the end-to-end test expected output.

fqazi replied:
Let me revert that and comment on this. It seemed unnecessary, but now it all makes sense.

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/common_relation.go:
```diff
@@ -83,4 +92,423 @@ func removeColumnTypeBackRefs(b BuildCtx, table catalog.TableDescriptor, id desc
 			})
 		}
 	}
+}*/
```
remove this comment block

postamar commented on pkg/sql/schemachanger/scop/ops.go:
```diff
@@ -28,6 +28,10 @@ type Ops interface {
+	// The type of stage doesn't matter for nil ops.
+	if len(ops) == 0 {
+		return nil
+	}
```
Why? It seems to me that if there are no ops, then we shouldn't be building a stage in the first place.

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_sequence.go:
```diff
@@ -38,39 +38,28 @@ func DropSequence(b BuildCtx, n *tree.DropSequence) {
+		}
 		if cascade != tree.DropCascade {
-			panic(pgerror.Newf(
+			return pgerror.Newf(
```
I find this new pattern where we return instead of panicking quite confusing. My first reaction was to think this was a bug because the return value of  `scpb.ForEachRelationDependedOnBy` is ignored. 

I had to look up the definition of `scpb.ForEachRelationDependedOnBy` to reassure myself that the error will indeed get panicked instead of being quietly ignored. This function and the others like it violate an unspoken style convention which is that `ForEach` functions bubble up the error value returned by the function argument.

Instead, why not have the anonymous fuction passed to `scpb.ForEachRelationDependedOnBy` and friends simply not have a return type at all? Seems more straightforward to me.

fqazi replied:
The intention here was that the iteration could be stopped with an error? @postamar Would a boolean return value make more sense for the call back?

fqazi replied:
Oh I take it back it's already pretty broken, makes more sense to drop the return value completely, since ForEachNode has no way of stopping the iteration

postamar reviewed (commented):
Thanks for doing this, this is good work. I only have minor complaints.

ajwerner commented on pkg/sql/schemachanger/scpb/element_uml_generator.go:
```diff
@@ -0,0 +1,84 @@
+		fmt.Fprintf(os.Stderr, "%v\n", err)
+		exit.WithCode(exit.FatalError())
+	}
+}
```
`cmd` is for cli tools that somebody might want to actually invoke. This pattern of defining code generators inline but build tagged so they don't generally run exists elsewhere. I think it's pretty elegant.

fqazi commented on pkg/sql/schemachanger/scplan/scopt/rules.go:
```diff
@@ -0,0 +1,53 @@
+}
+
+// TODO(fqazi): For create operations with the decomposition work
+// we need to implement optimizations combine edges together.
```
// TODO(fqazi): For create operations we will need to have the ability
// to have transformations that will combine transitions into a single
// stage for execution. For example, a newly CREATE TABLE will be represented
// by a TABLE, COLUMN, and INDEX elements (among others), all of the operations
// for these elements are executable in a single stage. Having them execute
// across multiple stages would be problematic both from a validation and
// correctness viewpoint.

fqazi commented on pkg/sql/schemachanger/scplan/scopt/rules.go:
```diff
@@ -0,0 +1,53 @@
+	relation, relationTarget, relationNode := targetNodeVars("rel")
+	var id rel.Var = "id"
+	registerNoOpEdges(screl.MustQuery(
+		relation.Type((*scpb.Table)(nil), (*scpb.View)(nil), (*scpb.Sequence)(nil), (*scpb.Type)(nil)),
```
Good catch, that shouldn't have been there.

fqazi commented on pkg/sql/schemachanger/scplan/scopt/registry.go:
```diff
@@ -0,0 +1,65 @@
+		}
+	}
+	return nil
+}
```
The graph is decorated, however, I think it makes sense to deep copy decorations/mutations and leaves the main structure as shallow copied.

fqazi commented on pkg/sql/schemachanger/scplan/plan.go:
```diff
@@ -77,7 +78,9 @@ func MakePlan(initial scpb.State, params Params) (_ Plan, err error) {
-
+	if err := scopt.OptimizePlan(g); err != nil {
+		return Plan{}, err
+	}
```
I want to separately rename these, or divide them into two categories as a refactor. The current things are transformations that are needed for *executable* plans. I think in later phases these same things can be used for actual optimizations to skip steps in the plan. For that phase, I think we should add a knob for disabling this stuff.

fqazi commented on pkg/sql/schemachanger/rel/database.go:
```diff
@@ -103,6 +103,43 @@ func (t *Database) Insert(v interface{}) error {
+			if err := t.Delete(v); err != nil {
+				return err
+			}
+		}
```
No clue what happened with reviewable, but there is no delete code any more only code for decoration.

fqazi commented on pkg/sql/schemachanger/rel/database.go:
```diff
@@ -103,6 +103,43 @@ func (t *Database) Insert(v interface{}) error {
+
+func (t *Database) delete(e *entity) error {
+	self := e.getComparableValue(t.schema, Self)
+	delete(t.entities, self)
```
No clue what happened with reviewable, but there is no delete code any more only code for decoration.

fqazi committed:
sql: create view does not add type/relation deps in an order

Previously, CREATE VIEW would add type or relation dependencies
in an arbitrary order. This was inadequate because it makes
certain types of deterministic tests very difficult, specifically
ones like the declarative schema changer end to end test. To
address this, this patch will order these lists based on
descriptor IDs.

Release note: None

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/common_relation.go:
```diff
@@ -17,70 +17,432 @@ import (
+	exprTypeComputed exprType = 2
+)
+
+// decomposeExprToTypeRef converts and inserts type references from
+// an expression.
+func decomposeExprToElements(
```
there's a few patterns in these decompositions I'm not sure about.

How much of the BuildCtx do we need? My sense is for most of it, not much. Fundamentally, I'm interested in asking whether we can define the decomposition without the tight coupling to the builder. I think it'll be valuable to perform full decompositions independently. I know it's a lot of packages but, what do you think about pulling the decomposition out into a package like `scelem`?

Second, the `addIfDuplicateDoesNotExistForDir` pattern smells to me.

ajwerner replied:
It'd be fun also to be able to test the decomposition stuff by starting with descriptors, decomposing them, and then running planning and execution to rebuild them and making sure that that still works.

postamar replied:
I'd be inclined to leave this code as it is and not worry about it, I'll be making a lot of changes to it soon anyway when working on the builder's view of the catalog state in terms of elements instead of descriptors.

ajwerner commented on pkg/sql/schemachanger/scgraph/graph.go:
```diff
@@ -110,6 +114,31 @@ func New(initial scpb.State) (*Graph, error) {
 	return &g, nil
 }
 
+// CloneForMutation shallow copies the main graph structure, and deep copies
+// any mutations / decorations on the graph.
+func (g *Graph) CloneForMutation() *Graph {
```
Can you use the word `Shallow` in the name of this method?

ajwerner commented on pkg/sql/schemachanger/scpb/element_uml_generator.go:
```diff
@@ -0,0 +1,84 @@
+		// The parent tag has a list of elements that are the parents
+		// to this element. We will collect these and emit them later
+		// in the PlantUML syntax.
+		if parentTag := elementProtoType.Field(i).Tag.Get("parent"); parentTag != "" {
```
nit: make a helper to turn the field into a string slice.

ajwerner commented on pkg/sql/schemachanger/scpb/elements_generated.go:
```diff
@@ -12,22 +12,338 @@
+
 func (e Column) element() {}
+
+// ForEachColumn iterates over nodes of type Column.
+func ForEachColumn (b NodeIterator, elementFunc func(element *Column) ) {
```
I don't see these used.

ajwerner commented on pkg/sql/schemachanger/scgraph/graph.go:
```diff
@@ -234,6 +263,18 @@ func (g *Graph) AddDepEdge(
 	return nil
 }
 
+// MarkOpEdgeAsNoOp marks an edge as no-op, so that no operations are emitted
+//during planning.
+func (g *Graph) MarkOpEdgeAsNoOp(edge *OpEdge) {
+	g.noOpEdges[edge] = true
+}
+
+// IsOpEdgeNoOp checks if an edge is marked as an edge that should emit no
+// operations.
+func (g *Graph) IsOpEdgeNoOp(edge *OpEdge) bool {
+	return g.noOpEdges[edge]
+}
```
nit: can we change this terminology to say something about the edge being optimized away as opposed to being a `NoOp`?

ajwerner commented on pkg/sql/schemachanger/scop/ops.go:
```diff
@@ -28,6 +28,10 @@ type Ops interface {
 // returns an implementation of Ops corresponding to that type. The set of ops
 // must all be the same type, otherwise MakeOps will panic.
 func MakeOps(ops ...Op) Ops {
+	// The type of stage doesn't matter for nil ops.
+	if len(ops) == 0 {
+		return nil
+	}
```
I really wish I understood why this happens better. Can we eliminate this in the planning layer and avoid making this call?

postamar replied:
I'd be inclined to leave this be and revisit empty stages and other optimization questions at a later point in time, if that's OK with you.

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72856)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/sql/schemachanger/scbuild/internal/scbuildstmt/common_relation.go, line 103 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72856#-MotRXb9rQNL6FFDUNnq-r12-103:-MpImW4S7b33871iE4Yy:b-csoq2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff8c8e8c6453ae12e7c43468e5da737d5378abd9/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/common_relation.go#L103)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I'd be inclined to leave this code as it is and not worry about it, I'll be making a lot of changes to it soon anyway when working on the builder's view of the catalog state in terms of elements instead of descriptors.
</blockquote></details>

I cleaned up the addIfDuplicateDoesNotExistDir and replaced it with explicit things for the only cases where it's needed. Relationship dependencies and cascaded drops
___
*[pkg/sql/schemachanger/scgraph/graph.go, line 119 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72856#-MogW497Ai5l-8upJIxj-r12-119:-MpImhmV3Ck85kLUCs9a:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff8c8e8c6453ae12e7c43468e5da737d5378abd9/pkg/sql/schemachanger/scgraph/graph.go#L119)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Can you use the word `Shallow` in the name of this method?
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scop/ops.go, line 34 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72856#-MoommfUeb_QFB0fvr2a-r12-34:-MpImkAtACoXAH3_ckAD:b-pjh5ft) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff8c8e8c6453ae12e7c43468e5da737d5378abd9/pkg/sql/schemachanger/scop/ops.go#L34)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I'd be inclined to leave this be and revisit empty stages and other optimization questions at a later point in time, if that's OK with you.
</blockquote></details>

I added an explicit TODO. I have WIP branch that does this after the stages are built, but will see if I can push it earlier without having a mess. I think its doable.
___
*[pkg/sql/schemachanger/scpb/elements_generated.go, line 23 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72856#-MogW496Zgtd342M6vyw-r12-23:-MpImtgu61kQ20J3KFUO:b-jy3oba) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff8c8e8c6453ae12e7c43468e5da737d5378abd9/pkg/sql/schemachanger/scpb/elements_generated.go#L23)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I don't see these used. 
</blockquote></details>

Done. 

Replaced ForEachNode calls
___
*[pkg/sql/schemachanger/rel/database.go, line 125 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72856#-MogW497Ai5l-8upJIxi-r5-125:-Mp8Ghfi627imyV0WHwp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/015442d7f612f0f817bfc1807849d31a62e22e50/pkg/sql/schemachanger/rel/database.go#L125)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

The same check is the first line of `Delete`, no need for the outer `if` here
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/rel/database.go, line 132 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72856#-MogW497Ai5l-8upJIxi-r5-132:-Mp8Gi-NDnRn3WKEA8_U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/015442d7f612f0f817bfc1807849d31a62e22e50/pkg/sql/schemachanger/rel/database.go#L132)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I have a modest preference to delete from the map after the indexes. I don't know why. Something feels better about it :/ 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
I'm happy when @postamar and CI are happy. The decomposition is good. We'll massage the other pieces in smaller follow-ups. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72856#-:-MpJ8bkr0ZtGFMLyc-48:b-oiy7zo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)



<!-- Sent from Reviewable.io -->

postamar reviewed (approved):
Let's merge this now.

fqazi commented:
bors r+
